pi5/pinctrl_utils.go

```diff
@@ -0,0 +1,287 @@
+	if err != nil {
+		return INVALID_ADDR, err
+	}
+
+	var addrRanges []rangeInfo
+	if physAddress == INVALID_ADDR { // Case 1: We are the Child Node. No addr has been set. Read the reg file to get the physical address within our parents space.
+		physAddress, err = getRegAddr(currNodePath, numPAddrCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+	} else { // Case 2: We use the ranges property to continue mapping a child addr into our parent addr space.
+		addrRanges, err = getRangesAddrInfo(currNodePath, numCAddrCells, numPAddrCells, numAddrSpaceCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+		// getRangesAddrInfo returns a list of all possible child address ranges our physical address can fall into. We must see which range to use, so that we can map our physical address into the correct parent address range.
+		for _, addrRange := range addrRanges {
+
+			if addrRange.childAddr <= physAddress && physAddress <= addrRange.childAddr+addrRange.addrSpaceSize {
+				physAddress -= addrRange.childAddr  // get the offset beween the address and child base address
+				physAddress += addrRange.parentAddr // now address has been mapped into parent space.
+				break
+			}
+		}
+	}
+
+	numCAddrCells = numPAddrCells
+	currNodePath = parentNodePath
+	return setGPIONodePhysAddrHelper(currNodePath, physAddress, numCAddrCells)
+}
+
+// Creates a virtual Page to access/manipulate memory related to gpiochip data
+func (b *pinctrlpi5) createGPIOVPage(memPath string) error {
+	var err error
+
+	// Open the 'file' you are trying to map. Note: /dev/gpiomem0 is an alias, so it's file length returns 0
+	fileFlags := os.O_RDWR | os.O_SYNC
+	b.memFile, err = os.OpenFile("/dev/gpiomem0", fileFlags, 0666) // 0666 is an octal representation of: file is readable / writeable by anyone
+	if err != nil {
+		return fmt.Errorf("failed to open %s: %w\n", memPath, err)
+	}
+
+	pageSize := uint64(syscall.Getpagesize())
+	pageOffset := b.physAddr & (pageSize - 1) // difference between base address of the page and the address we're actually tring to access
+	//pageStartAddr := int64(b.physAddr - pageOffset) // base address of the page. needed when we can't open dev/gpiomem0
+	lenMapping := int(pageOffset) + int(b.chipSize) // total amount of memory needed to be mapped. + pageoffset is because we're starting from the base address of the page, not our actual physical address
+
+	vPage, err := mmap.MapRegion(b.memFile, lenMapping, mmap.RDWR, 0, 0) // 0 flag = shared, 0 offset because we are starting from base address pointing to /gpiomem0
+	if err != nil {
+		if err == syscall.ENOMEM {
+			return fmt.Errorf("failed to mmap: cannot allocate memory for mmap call %w", err)
+		}
+		return fmt.Errorf("failed to mmap: %w\n", err)
+	}
+
+	tempVirtAddr := unsafe.Pointer(&vPage[0]) // we can't directly assign something of type Pointer to *uint64 in one line of code
+	b.virtAddr = (*uint64)(tempVirtAddr)
```

Comment:
where do we use the virtual address?

Reply:
we will use it once we start accessing memory in the gpio banks to flip bits. This address, in addition to the offset for each respective gpio bank, serves as the "base" address of a gpio bank. 

From there, if we want to access a specific pin (let's say #19 for example, which is in gpio bank 0) we would access gpiopins[19] by adding 

               "base"          +         (19 *size of (pin)) 
= virt.Addr + bank0 offset + (19 *size of (pin)) 

does that answer your question?

Reply:
that makes sense thanks for explaining! Could we make `b.VirtAddr `a `*mmap.MMap` instead of a uint64? that way its a pointer but we dont need to directly store the address

Reply:
I tried to store a pointer but the type converstion isn't supported . I think it might just be easier if I store the vPage. it would be a byte Slice just stored in the struct for a board

Reply:
what type conversion isnt supported? you should just create to mmap object then do b.virtAddr = &mmap